File: src/main/java/cz/vhromada/catalog/entity/Season.java
package cz.vhromada.catalog.entity;

import java.util.List;
import java.util.Objects;

import cz.vhromada.catalog.common.Language;
import cz.vhromada.catalog.common.Movable;

/**
 * A class represents season.
 *
 * @author Vladimir Hromada
 */
public class Season implements Movable {

    /**
     * SerialVersionUID
     */
    private static final long serialVersionUID = 1L;

    /**
     * ID
     */
    private Integer id;

    /**
     * Number of season
     */
    private int number;

    /**
     * Starting year
     */
    private int startYear;

    /**
     * Ending year
     */
    private int endYear;

    /**
     * Language
     */
    private Language language;

    /**
     * Subtitles
     */
    private List<Language> subtitles;

    /**
     * Note
     */
    private String note;

    /**
     * Position
     */
    private int position;

    @Override
    public Integer getId() {
        return id;
    }

    @Override
    public void setId(final Integer id) {
        this.id = id;
    }

    /**
     * Returns number of season.
     *
     * @return number of season
     */
    public int getNumber() {
        return number;
    }

    /**
     * Sets a new value to number of season.
     *
     * @param number new value
     */
    public void setNumber(final int number) {
        this.number = number;
    }

    /**
     * Returns starting year.
     *
     * @return starting year
     */
    public int getStartYear() {
        return startYear;
    }

    /**
     * Sets a new value to starting year.
     *
     * @param startYear new value
     */
    public void setStartYear(final int startYear) {
        this.startYear = startYear;
    }

    /**
     * Returns ending year.
     *
     * @return ending year
     */
    public int getEndYear() {
        return endYear;
    }

    /**
     * Sets a new value to ending year.
     *
     * @param endYear new value
     */
    public void setEndYear(final int endYear) {
        this.endYear = endYear;
    }

    /**
     * Returns language.
     *
     * @return language
     */
    public Language getLanguage() {
        return language;
    }

    /**
     * Sets a new value to language.
     *
     * @param language new value
     * @throws IllegalArgumentException if new value is null
     */
    public void setLanguage(final Language language) {
        this.language = language;
    }

    /**
     * Returns subtitles.
     *
     * @return subtitles
     */
    public List<Language> getSubtitles() {
        return subtitles;
    }

    /**
     * Sets a new value to subtitles.
     *
     * @param subtitles new value
     */
    public void setSubtitles(final List<Language> subtitles) {
        this.subtitles = subtitles;
    }

    /**
     * Returns note.
     *
     * @return note
     */
    public String getNote() {
        return note;
    }

    /**
     * Sets a new value to note.
     *
     * @param note new value
     */
    public void setNote(final String note) {
        this.note = note;
    }

    @Override
    public int getPosition() {
        return position;
    }

    @Override
    public void setPosition(final int position) {
        this.position = position;
    }

    @Override
    public boolean equals(final Object obj) {
        if (this == obj) {
            return true;
        }

        if (!(obj instanceof Season) || id == null) {
            return false;
        }

        return id.equals(((Season) obj).id);
    }

    @Override
    public int hashCode() {
        return Objects.hashCode(id);
    }

    @Override
    public String toString() {
        return String.format("Season [id=%d, number=%d, startYear=%d, endYear=%d, language=%s, subtitles=%s, note=%s, position=%d]", id, number, startYear,
                endYear, language, subtitles, note, position);
    }

}

File: catalog-core/src/main/kotlin/com/github/vhromada/catalog/domain/Season.kt
package com.github.vhromada.catalog.domain

import com.github.vhromada.common.domain.Audit
import com.github.vhromada.common.entity.Language
import com.github.vhromada.common.entity.Movable
import com.github.vhromada.common.utils.sorted
import org.hibernate.annotations.Fetch
import org.hibernate.annotations.FetchMode
import java.util.Objects
import javax.persistence.CascadeType
import javax.persistence.CollectionTable
import javax.persistence.Column
import javax.persistence.ElementCollection
import javax.persistence.Entity
import javax.persistence.EnumType
import javax.persistence.Enumerated
import javax.persistence.FetchType
import javax.persistence.GeneratedValue
import javax.persistence.GenerationType
import javax.persistence.Id
import javax.persistence.JoinColumn
import javax.persistence.ManyToOne
import javax.persistence.OneToMany
import javax.persistence.OrderBy
import javax.persistence.SequenceGenerator
import javax.persistence.Table

/**
 * A class represents season.
 *
 * @author Vladimir Hromada
 */
@Entity
@Table(name = "seasons")
data class Season(
    /**
     * ID
     */
    @Id
    @SequenceGenerator(name = "season_generator", sequenceName = "seasons_sq", allocationSize = 1)
    @GeneratedValue(strategy = GenerationType.SEQUENCE, generator = "season_generator")
    override var id: Int?,

    /**
     * Number of season
     */
    @Column(name = "season_number")
    val number: Int,

    /**
     * Starting year
     */
    @Column(name = "start_year")
    val startYear: Int,

    /**
     * Ending year
     */
    @Column(name = "end_year")
    val endYear: Int,

    /**
     * Language
     */
    @Column(name = "season_language")
    @Enumerated(EnumType.STRING)
    val language: Language,

    /**
     * Subtitles
     */
    @ElementCollection(fetch = FetchType.EAGER)
    @CollectionTable(name = "season_subtitles", joinColumns = [JoinColumn(name = "season")])
    @Enumerated(EnumType.STRING)
    @Fetch(FetchMode.SELECT)
    val subtitles: List<Language>,

    /**
     * Note
     */
    val note: String?,

    /**
     * Position
     */
    override var position: Int?,

    /**
     * Episodes
     */
    @OneToMany(mappedBy = "season", fetch = FetchType.EAGER, cascade = [CascadeType.ALL], orphanRemoval = true)
    @OrderBy("position, id")
    @Fetch(FetchMode.SELECT)
    val episodes: MutableList<Episode>
) : Audit(), Movable {

    /**
     * Show
     */
    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "tv_show")
    var show: Show? = null

    override fun updatePosition(position: Int) {
        super.updatePosition(position)

        for (i in episodes.sorted().indices) {
            episodes[i].updatePosition(i)
        }
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) {
            return true
        }
        return if (other !is Season || id == null) {
            false
        } else {
            id == other.id
        }
    }

    override fun hashCode(): Int {
        return Objects.hashCode(id)
    }

}
